### PR TITLE
simplify creation of CMake files for Python tests

### PR DESCRIPTION
**Pull request by 3nids**
this simplifies the cmake macro by adding a template file and using `configure_file`.

### CODE REVIEW

File: cmake_templates/PyQgsTest.cmake.in [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+SET(ENV{QGIS_PREFIX_PATH} "@QGIS_PYTEST_PREFIX_PATH@")
```
Comment (m-kuhn) [inline cmake_templates/PyQgsTest.cmake.in:2]: Is there a reason to keep commands uppercase in here?

Comment (3nids) [reply cmake_templates/PyQgsTest.cmake.in]: thanks, fixed

File: cmake/UsePythonTest.cmake [hunk 2]
```diff
@@ -20,96 +19,59 @@
...
-")
-  ENDIF(WIN32)
 
-  SET(TEST_TIMEOUT 0)
+  set(_QGIS_PYTEST_PYTHONPATHS "${PYTHON_OUTPUT_DIRECTORY}" "${PYTHON_OUTPUT_DIRECTORY}/plugins" "${CMAKE_SOURCE_DIR}/tests/src/python")
 
-  FOREACH(_in ${ARGN})
-    STRING(REGEX MATCH "^([^=]+)=(.*)$" _out ${_in})
+  if(WIN32)
+    set(QGIS_PYTEST_SEPARATOR ";")
```
Comment (m-kuhn) [inline cmake/UsePythonTest.cmake:36]: Where is this used?

Could we use the cmake function `cmake_path(CONVERT <input> TO_NATIVE_PATH_LIST <out-var>)` instead

File: cmake/UsePythonTest.cmake [hunk 3]
```diff
@@ -20,96 +19,59 @@
...
+      set(QGIS_PYTEST_PREFIX_PATH "${QGIS_OUTPUT_DIRECTORY}/bin/${CMAKE_BUILD_TYPE}")
+      set(QGIS_PYTEST_LIBRARY_PATH "${QGIS_OUTPUT_DIRECTORY}/bin/${CMAKE_BUILD_TYPE};")
+      set(QGIS_PYTEST_PYTHONPATH "${_QGIS_PYTEST_PYTHONPATHS};")
+    endif()
+  else()
+    set(QGIS_PYTEST_SEPARATOR ":")
+    set(QGIS_PYTEST_PATH_VAR_NAME "LD_LIBRARY_PATH")
+    set(QGIS_PYTEST_PREFIX_PATH "${QGIS_OUTPUT_DIRECTORY}")
+    set(QGIS_PYTEST_LIBRARY_PATH "${QGIS_PYTEST_PYENV}${QGIS_OUTPUT_DIRECTORY}/lib")
+    string(REPLACE ";" ":" QGIS_PYTEST_PYTHONPATH "${_QGIS_PYTEST_PYTHONPATHS};")
```
Comment (m-kuhn) [inline cmake/UsePythonTest.cmake:52]: Could this be managed as cmake list and then sent through `cmake_path(CONVERT <input> TO_NATIVE_PATH_LIST <out-var>)` (on all platforms)?

File: cmake/UsePythonTest.cmake [hunk 4]
```diff
@@ -20,96 +19,59 @@
...
+  if(WIN32)
+    set(QGIS_PYTEST_SEPARATOR ";")
+    set(PATH_VAR_NAME "PATH")
+    if(USING_NINJA OR USING_NMAKE)
+      set(QGIS_PYTEST_PREFIX_PATH "${QGIS_OUTPUT_DIRECTORY}/bin")
+      set(QGIS_PYTEST_LIBRARY_PATH "${QGIS_OUTPUT_DIRECTORY}/bin;")
+      set(QGIS_PYTEST_PYTHONPATH "${_QGIS_PYTEST_PYTHONPATHS};")
+    else()
+      set(QGIS_PYTEST_PREFIX_PATH "${QGIS_OUTPUT_DIRECTORY}/bin/${CMAKE_BUILD_TYPE}")
+      set(QGIS_PYTEST_LIBRARY_PATH "${QGIS_OUTPUT_DIRECTORY}/bin/${CMAKE_BUILD_TYPE};")
```
Comment (m-kuhn) [inline cmake/UsePythonTest.cmake:44]:
```suggestion
      set(QGIS_PYTEST_PREFIX_PATH "${QGIS_OUTPUT_DIRECTORY}/bin/$<CONFIG>")
      set(QGIS_PYTEST_LIBRARY_PATH "${QGIS_OUTPUT_DIRECTORY}/bin/$<CONFIG>;")
```

Not an expert on the mather, but IIRC the MSVC generator is a multibuild generator and we cannot trust the `CMAKE_BUILD_TYPE` to be correct (or to be set at all)

File: cmake/UsePythonTest.cmake [hunk 5]
```diff
@@ -20,96 +19,59 @@
...
 
-  SET(TEST_TIMEOUT 0)
+  set(_QGIS_PYTEST_PYTHONPATHS "${PYTHON_OUTPUT_DIRECTORY}" "${PYTHON_OUTPUT_DIRECTORY}/plugins" "${CMAKE_SOURCE_DIR}/tests/src/python")
 
-  FOREACH(_in ${ARGN})
-    STRING(REGEX MATCH "^([^=]+)=(.*)$" _out ${_in})
+  if(WIN32)
+    set(QGIS_PYTEST_SEPARATOR ";")
+    set(PATH_VAR_NAME "PATH")
+    if(USING_NINJA OR USING_NMAKE)
```
Comment (m-kuhn) [inline cmake/UsePythonTest.cmake:38]:
```suggestion
    if(NOT CMAKE_CONFIGURATION_TYPES)
```

This should be more appropriate?

File: cmake/UsePythonTest.cmake [hunk 6]
```diff
@@ -1,115 +1,82 @@
...
-SET(ENV{QGIS_PREFIX_PATH} \"${QGIS_OUTPUT_DIRECTORY}\")
-SET(ENV{LD_LIBRARY_PATH} \"${pyenv}${QGIS_OUTPUT_DIRECTORY}/lib:\$ENV{LD_LIBRARY_PATH}\")
-SET(ENV{PYTHONPATH} \"${PYTHON_OUTPUT_DIRECTORY}:${PYTHON_OUTPUT_DIRECTORY}/plugins:${CMAKE_SOURCE_DIR}/tests/src/python:\$ENV{PYTHONPATH}\")
-MESSAGE(\"export LD_LIBRARY_PATH=\$ENV{LD_LIBRARY_PATH}\")
-")
-  ENDIF(WIN32)
 
-  SET(TEST_TIMEOUT 0)
+  set(_QGIS_PYTEST_PYTHONPATHS "${PYTHON_OUTPUT_DIRECTORY}" "${PYTHON_OUTPUT_DIRECTORY}/plugins" "${CMAKE_SOURCE_DIR}/tests/src/python")
+  cmake_path(CONVERT "${_QGIS_PYTEST_PYTHONPATHS}" TO_NATIVE_PATH_LIST QGIS_PYTEST_PYTHONPATH)
```
Comment (m-kuhn) [inline cmake/UsePythonTest.cmake:35]: Sorry, that was only added with cmake 3.20, probably too new

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**